[5.2] Resolve parameter fields with asterisks when validating arrays

sebdesign opened the pull request:
When validating arrays with rules that accept other fields as parameters, like `required_*`, the field names should reflect the structure of the array we are validating against. E.g.:

``` php
'person.*.first_name' => 'required_with:last_name'
```

This rule would **NOT** check if `last_name` is present in _each_ person, but if it is present in the root array data. While this is a useful feature, we need a way to accept fields as parameters which are scoped to the current item when we perform array validation.

The most natural way would be to specify the fields with `*` character like we do in the rule keys. Unfortunately this doesn't work in 5.2, so this PR addresses and fixes that issue. It also works on multidimentional (multiple `*`) array structures. Here is an example:

``` php
'person.*.first_name' => 'required_with:person.*.last_name'
```

There is also the issue of the missing array keys, that should be addressed in #11885 .
For now all my tests are specifying `'field' => null` when we need to check for a field, but this should be fixed to work even if the needed field is missing from the data.

sebdesign committed:
Resolve parameter fields with asterisks when validating arrays

GrahamCampbell commented on src/Illuminate/Validation/Validator.php:
```diff
@@ -515,6 +515,57 @@ protected function addError($attribute, $rule, $parameters)
+     * Replace each field which has asterisks with the numeric keys of the given attribute.
+     *
+     * @param  array  $fields
+     * @param  string $attribute
```
cs

GrahamCampbell commented on src/Illuminate/Validation/Validator.php:
```diff
@@ -515,6 +515,57 @@ protected function addError($attribute, $rule, $parameters)
+     *
+     * E.g. 'foo.*.bar.*.baz', [1, 2] -> foo.1.bar.2.baz
+     *
+     * @param  strint $field
```
cs

sebdesign committed:
Fix cs

sebdesign commented:
@GrahamCampbell done

garygreen commented:
+1

tomswinkels commented:
+1

GrahamCampbell commented:
Lots of duplicate code here. ;(

sebdesign commented:
@GrahamCampbell could you be more specific please? I know the tests are not DRY at all, I can refactor them I guess. How about the implementation code?

garygreen commented:
@sebdesign maybe in the `validate()` method it can decide what rules can have dot-format attributes and pass thru the processed parameters, maybe `->parseParameters()` or something. That will get rid of the need for `$parameters = $this->replaceParameterFields($parameters, $attribute);` in a lot of rules.

sebdesign commented:
Thank you for your feedback @GrahamCampbell! I will revise my code as soon
as possible.
On Feb 3, 2016 1:07 PM, "Gary Green" <notifications@github.com> wrote:

> @sebdesign <https://github.com/sebdesign> maybe in the validate() method
> it can decide what rules can have dot-format attributes and pass thru the
> processed parameters, maybe ->parseParameters() or something. That will
> get rid of the need for $parameters =
> $this->replaceParameterFields($parameters, $attribute); in a lot of rules.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/framework/pull/11927#issuecomment-179169309>.
>

sebdesign committed:
Refactor and prevent some potential regressions

Move some methods around, improve naming, and remove repeated method calls,
because elegant code is good code.

Also, determine which rules need their parameters fields normalized.
For example, we don't want to replace asterisks with numbers on regex rules.

sebdesign commented:
@GrahamCampbell @taylorotwell @garygreen I've refactored! No more ugly copy-pasta.

garygreen commented:
MUCH better, good work ! :smile:

sebdesign commented:
@garygreen thank you! I think the TaylorOtwell™ style comment was the hardest part. :)

KoenRijpstra commented:
+1

lucasmichot commented on src/Illuminate/Validation/Validator.php:
```diff
@@ -2262,6 +2279,64 @@ protected function normalizeRule($rule)
+        if (preg_match_all('/\.(\d+)\./', $attribute, $keys)) {
+            array_shift($keys);
+
+            return array_collapse($keys);
```
Calling helpers should be avoided.
`return array_collapse($keys);`

sebdesign replied:
@lucasmichot fixed.

lucasmichot replied:
:+1:

vlakoff commented on src/Illuminate/Validation/Validator.php:
```diff
@@ -2262,6 +2279,64 @@ protected function normalizeRule($rule)
+        }
+
+        return [];
+    }
```
You can simplify as so:
```
if (preg_match_all('/\.(\d+)\./', $attribute, $keys)) {
    return $keys[1];
}

return [];
```
In fact, `$keys[1]` is set even if there is no match, so you could even do:
```
preg_match_all('/\.(\d+)\./', $attribute, $keys);

return $keys[1];
```

lucasmichot replied:
:+1:

sebdesign replied:
@vlakoff Thanks! Are you sure there's no need to collapse the `$keys[1]` though?

vlakoff replied:
`$keys` is an array of arrays, you removed the unwanted `$keys[0]` then collapsed the remaining of `$keys` (so, `$keys[1]`) to get a flat array.

Here, I directly pick `$keys[1]` which is already a flat array. You don't have to (and cannot) collapse it.

sebdesign replied:
@vlakoff thanks to your suggestions, I have committed the changes you proposed. :+1:

sebdesign committed:
Simplify getNumericKeys

taylorotwell commented:
Looks like this needs re-basing? Did you want to make any changes to the tests as you mentioned in the original PR?

GrahamCampbell commented:
Please squash to one commit and fix the tests.

prafful-panwar commented:
**Validation Rule :**
'event' => 'required|array',
'event.*.start_date_time' => 'required|date_format:Y-m-d|after:today',
'event.*.end_date_time' => 'required|date_format:Y-m-d|after:today',
'event.*.description' => 'required|spam_free',

**Error Message**
if ($messages->has('event')) {
    $fieldId = 'event';
    $errors [] = getErrorMessage($fieldId, $messages->first($fieldId));
 }
if ($messages->has('event.*.start_date_time')) {
     $fieldId = 'event.*.start_date_time';
     $errors [] = getErrorMessage($fieldId, $messages->first($fieldId));
}
if ($messages->has('event.*.end_date_time')) {
      $fieldId = 'event.*.end_date_time';
      $errors [] = getErrorMessage($fieldId, $messages->first($fieldId));
}
if ($messages->has('event.*.description')) {
      $fieldId = 'event.*.description';
      $errors [] = getErrorMessage($fieldId, $messages->first($fieldId));
}
**Validation Message**
{
      "fieldld": "event.*.start_date_time",
      "message": "The event.0.start_date_time field is required."
},
{
      "fieldld": "event.*.end_date_time",
      "message": "The event.0.end_date_time field is required."
}, 
{
       "fieldld": "event.*.description",
       "message": "The event.0.description field is required."
},

Can anyone help me why its not validation array
